utils/ofi_tcp_keepalive.patch

```diff
@@ -0,0 +1,139 @@
+ 	ofi_io_uring_t ring;
+@@ -774,8 +777,8 @@ do {						\
+ 	ofi_straddr(_src_buf, &_src_len, ofi_translate_addr_format(ofi_sa_family(_src_addr)),	\
+ 		    _src_addr);									\
+ 												\
+-	FI_WARN(&xnet_prov, FI_LOG_EP_CTRL, "error on %s From %s : %s\n", _buf, _src_buf,	\
+-		fi_strerror(err));								\
++	FI_WARN(&xnet_prov, FI_LOG_EP_CTRL, "ep %p state %u error on %s From %s : %s\n",	\
++		ep, ep->state, _buf, _src_buf, fi_strerror(err));				\
+ } while(0)
+ 
+ #endif //_XNET_H_
+diff --git a/prov/tcp/src/xnet_cm.c b/prov/tcp/src/xnet_cm.c
+index 1ae35d27d..64d66922a 100644
+--- a/prov/tcp/src/xnet_cm.c
++++ b/prov/tcp/src/xnet_cm.c
+@@ -203,6 +203,8 @@ void xnet_req_done(struct xnet_ep *ep)
+ 	ret = xnet_req_done_internal(ep);
+ 	if (ret)
+ 		goto disable;
++
++	xnet_disable_keepalive(ep);
+ 	return;
+ 
+ disable:
+diff --git a/prov/tcp/src/xnet_ep.c b/prov/tcp/src/xnet_ep.c
+index 3fc566b7d..9f26d9c63 100644
+--- a/prov/tcp/src/xnet_ep.c
++++ b/prov/tcp/src/xnet_ep.c
+@@ -156,6 +156,66 @@ static void xnet_set_no_port(SOCKET sock)
+ #define xnet_set_no_port(sock)
+ #endif
+ 
++int
++xnet_disable_keepalive(struct xnet_ep *ep)
++{
++	int optval = 0;
++	int ret;
++
++	ret = setsockopt(ep->bsock.sock, SOL_SOCKET, SO_KEEPALIVE, (char *)&optval,
++			 sizeof(optval));
++	if (ret) {
++		FI_WARN(&xnet_prov, FI_LOG_EP_CTRL, "set SO_KEEPALIVE failed %d", ret);
++		return -ofi_sockerr();
++	}
++
++	FI_INFO(&xnet_prov, FI_LOG_EP_CTRL, "ep %p KEEPALIVE is disabled.\n", ep);
++	return ret;
++}
++
++int
++xnet_enable_keepalive(struct xnet_ep *ep)
++{
++	int optval = 1;
++	int idle_time = 5;
++	int keep_intvl = 2;
++	int keep_cnt = 2;
++	int ret;
++
++	ret = setsockopt(ep->bsock.sock, SOL_SOCKET, SO_KEEPALIVE, (char *)&optval,
++			 sizeof(optval));
++	if (ret) {
++		FI_WARN(&xnet_prov, FI_LOG_EP_CTRL, "set SO_KEEPALIVE failed %d", ret);
++		return -ofi_sockerr();
++	}
++
++	ret = setsockopt(ep->bsock.sock, IPPROTO_TCP, TCP_KEEPIDLE, (char *)&idle_time,
++			 sizeof(idle_time));
++	if (ret) {
++		FI_WARN(&xnet_prov, FI_LOG_EP_CTRL, "set TCP_KEEPIDLE failed %d", ret);
++		return -ofi_sockerr();
++	}
++
++	ret = setsockopt(ep->bsock.sock, IPPROTO_TCP, TCP_KEEPINTVL, (char *)&keep_intvl,
++			 sizeof(keep_intvl));
++	if (ret) {
++		FI_WARN(&xnet_prov, FI_LOG_EP_CTRL, "set TCP_KEEPINTVL failed %d", ret);
++		return -ofi_sockerr();
++	}
++
++	ret = setsockopt(ep->bsock.sock, IPPROTO_TCP, TCP_KEEPCNT, (char *)&keep_cnt,
++			 sizeof(keep_cnt));
++	if (ret) {
++		FI_WARN(&xnet_prov, FI_LOG_EP_CTRL, "set SO_KEEPALIVE failed %d", ret);
++		return -ofi_sockerr();
++	}
++
++	FI_INFO(&xnet_prov, FI_LOG_EP_CTRL, "%p KEEPALIVE idle %d intvl %d cnt %d\n",
++		ep, idle_time, keep_intvl, keep_cnt);
++
++	return ret;
++}
++
+ int xnet_setup_socket(SOCKET sock, struct fi_info *info)
+ {
+ 	int ret, optval = 1;
+@@ -231,6 +291,8 @@ static int xnet_ep_connect(struct fid_ep *ep_fid, const void *addr,
+ 	if (!ep->addr)
+ 		return -FI_ENOMEM;
+ 
++	xnet_enable_keepalive(ep);
```

Comment:
Do you want to process error here?

btw, I tend to move this to the function `xnet_connect_done()`.

Reply:
can you please take care of this?

Reply:
I am not sure if keepalive timer would mess up with connecting timeout; I am too lazy to check the actual implementation, lol

Reply:
Hmm, I thought it should still proceed to connect even if set keepalive failed. And also connection will be kept retrying, and doubling enable keepalive should be fine, so it seems fine to only disable it after the connection is setup.

Though I might miss some cases, what is your concern here? Thanks.

Reply:
it's more about uncertainties since I don't know how this works, like how keepalive behaves if the connection is not created yet. For example, if there are some hiccups in the network that makes connecting take much longer, and then keepalive mechanism kicks in, etc...